### DISCUSSION TITLE
Configuring Audit.NET and getting user details from ASP.NET CORE app.

### DISCUSSION THREAD

**Discussion by tessierp** [thepirat000/Audit.NET#368] [Q&A]
Hi,

Consider the following code

            Instance._assemblyName = assemblyFullName;

            Configuration.Setup()
                .UseEntityFramework(_ => _
                    .AuditTypeMapper(t => typeof(AuditLog))
                    .AuditEntityAction<AuditLog>((ev, entry, entity) =>
                    {
                        var assembly = Assembly.Load(Instance._assemblyName);
                        var assemblyName = assembly.GetName();

                        entity.AuditId = Guid.NewGuid();
                        entity.Timestamp = DateTimeUtilities.GetUnixTimeSeconds(DateTime.UtcNow);
                        entity.TableName = entry.EntityType.Name;
                        entity.OperationData = entry.ToJson();
                        entity.Operation = entry.Action;
                        entity.OperationPerformedBy = getUserName();
                        entity.NameOfAssembly = assemblyName.Name;
                        entity.VersionOfAssembly = assemblyName.Version.ToString();
                    })
                .IgnoreMatchedProperties(true));`
`
Where getUserName is a delegate function that gets the current username. In an ASP.NET CORE application, how would you go about getting the username or details? Lets say you want to get the access_token to extract what you need. The only way to do that is by accessing the HttpContext. However, the context is not accessible from the Startup and I tried to use IHttpContextAccessor which does not work. I have seen some using the Thread's Principal to store the name which I guess would work but this is not ideal as I would be storing an access_token in the Name property of Identity. 

Any ideas how to best solve this?

**Comment by thepirat000** [comment]
You actually can access the httpcontext from the delegate. Take a look a [this](https://github.com/thepirat000/Audit.NET/issues/299#issuecomment-636412816) for example.

Why it didn't work in your case?

**Reply by tessierp** [reply]
IHttpContextAccessor gives me a HttpContext that is null every time. I did add the proper service for it, services.AddHttpContextAccessor().

**Reply by thepirat000** [reply]
Could you share the code you tried and didn't work, and why it didn't work?

**Reply by tessierp** [reply]
I did as per the example you sent. Basically same code. IHttpContextAccessor returns no context. I added the service, then use it and what is sent to me is a IHttpContextAccess with a null HttpContext. Did you try with .NET CORE 5.0? Perhaps I should go back to .NET CORE 3.1 and try.

**Reply by thepirat000** [reply]
Yes I did try with .net core 5.0. Here is what I did:

1- Installed the [audit web api template](https://github.com/thepirat000/Audit.NET/blob/master/src/Audit.WebApi/README.md#web-api-template-dotnet-new) with command:

`dotnet new -i Audit.webapi.template`

2- Created a new project:

`dotnet new webapiaudit -E -S`

3- Changed the target framework to be net5.0 on the csproj:

`<TargetFramework>net5.0</TargetFramework>`

And when I run the project, I can see the HttpContext:

![image](https://user-images.githubusercontent.com/9836380/107466047-19771d80-6b29-11eb-9cb1-f3c46ffea4e3.png)

**Reply by tessierp** [reply]
OK so you are using net5.0 where as I am using netcoreapp5.0. So the difference probably lies there.

**Reply by thepirat000** [reply]
`netcoreapp5.0` is not a valid moniker (check [this](https://docs.microsoft.com/en-us/dotnet/standard/frameworks)), latest is actually netcoreapp3.1. That must be the problem

**Reply by tessierp** [reply]
Alright, I found this doc...https://docs.microsoft.com/en-us/aspnet/core/migration/31-to-50?view=aspnetcore-5.0&tabs=visual-studio

So perhaps it is net5.0.. Now why would Microsoft allow me to compile with netcoreapp5.0 is beyond me. I'll try it tomorrow and see if that changes anything.

**Reply by tessierp** [reply]
Hello,

Did the appropriate changes, still can't get the HttpContext out of HttpContextAccessor...

![image](https://user-images.githubusercontent.com/6784384/107514481-b9f42e80-6b77-11eb-83d2-8b8db7bfadc9.png)

Application aside, it is your basic ASP.NET CORE WebAPI application... Not really sure how it is working for you while it is sending me a null value on my side.

**Reply by tessierp** [reply]
It could be that at the time I am asking for IHttpContextAccessor the HttpContext doesn't exist. It is not obvious from where in your code you are calling UseAuditCorrelationId but I assumed the Configure function which is from where I am calling it...

**Reply by tessierp** [reply]
I did some more testing and I am able to get a HttpContextAccessor in a controller that is valid but not during the applications Configuration state in Startup. I am not sure how you were able to achieve that.

I payed a closer attention at the code snapshot your provided above, you are executing your code in a Middleware! Now it makes sense to me why you get a valid HttpContextAccessor which is very different from the code here https://github.com/thepirat000/Audit.NET/issues/299  which is done in the Startup which wouldn't have any HttpContext ready at that point. I am not sure if others are doing it this way and if they are not sure how they got it to work.

So you build your own custom Middleware for that part I presume?

**Reply by tessierp** [reply]
Middleware was the solution to get IHttpContextAccessor. Again beats me how you manage to get a valid HttpContext from the Startup. Problem solved.

**Reply by thepirat000** [reply]
Glad it's solved but I don't get what you say about the middleware. It should not be the problem. Actually is not running **at** the startup. At the startup there is no httpcontext. The code *is* on the startup, but the custom action delegate is called on the context of the audit. So if you are calling DbContext.SaveChanges from an user action, the httpcontext should be accessible on the custom action delegate.

Here is the sample project: [issuenet5.zip](https://github.com/thepirat000/Audit.NET/files/5960141/issuenet5.zip)

**Reply by tessierp** [reply]
Yeah I know that which was strange to see in the example of the other post. Basically what I do is, where there is a create, update or delete, I call a Custom Action before, and setup is done once somewhere.

**Reply by CannyGent** [reply]
Hi - just to say I configured a Blazor app in latest .NET and successfully used the HttpContextAccessor method
